File: src/types/tracking.ts
export interface GPSLocation {
  latitude: number;
  longitude: number;
  heading?: number | null;
  speed?: number | null;
  accuracy?: number | null;
  altitude?: number | null;
  timestamp: number;
}

export interface DriverLocationRecord {
  id: string;
  driver_id: string;
  ride_id?: string | null;
  latitude: number;
  longitude: number;
  heading?: number | null;
  speed?: number | null;
  accuracy?: number | null;
  recorded_at: string;
  updated_at: string;
}

export type TrackingStatus = 'idle' | 'locating' | 'tracking' | 'error' | 'stopped';
export type PermissionStatus = 'prompt' | 'granted' | 'denied' | 'unsupported';

export interface TrackingState {
  driverLocation: GPSLocation | null;
  passengerLocation: GPSLocation | null;
  status: TrackingStatus;
  permissionStatus: PermissionStatus;
  gpsError: string | null;
  lastSyncTime: number | null;
  autoFollow: boolean;
}

// ─── Sprint 10C: Ride Tracking Session Types ─────────────────────────────────

export type TrackingSessionStatus =
  | 'Preparing'
  | 'DriverEnRoute'
  | 'PassengerPickup'
  | 'RideInProgress'
  | 'DestinationApproaching'
  | 'Completed'
  | 'Cancelled';

export interface TrackingSession {
  id: string;
  ride_id: string;
  driver_id: string;
  started_at: string;
  ended_at?: string;
  current_status: TrackingSessionStatus;
  last_location_at?: string;
  current_eta?: string;
  eta_minutes?: number;
  total_distance_km?: number;
  remaining_distance_km?: number;
  progress_percent?: number;
}

export interface ETAData {
  ride_id: string;
  eta_minutes?: number;
  current_eta_iso?: string;
  remaining_distance_km?: number;
  progress_percent?: number;
  current_status?: string;
  is_delayed?: boolean;
}

export interface LiveLocationUpdate {
  ride_id: string;
  driver_id: string;
  latitude: number;
  longitude: number;
  heading?: number;
  speed?: number;
  accuracy?: number;
  recorded_at: string;
}

export interface RouteProgress {
  ride_id: string;
  progress_percent: number;
  distance_traveled_km: number;
  remaining_distance_km: number;
  current_phase: string;
}

File: backend/app/schemas/tracking.py
from datetime import datetime
from typing import Optional
from uuid import UUID
from pydantic import BaseModel, Field, ConfigDict

from app.schemas.enums import TrackingStatus


class TrackingSessionResponse(BaseModel):
    """Public representation of an active RideTrackingSession."""

    id: UUID
    ride_id: UUID
    driver_id: UUID
    started_at: datetime
    ended_at: Optional[datetime] = None
    current_status: TrackingStatus
    last_location_at: Optional[datetime] = None
    current_eta: Optional[datetime] = None
    eta_minutes: Optional[float] = None
    total_distance_km: Optional[float] = None
    remaining_distance_km: Optional[float] = None
    progress_percent: Optional[float] = None

    model_config = ConfigDict(from_attributes=True)


class LocationUpdateEvent(BaseModel):
    """Payload for a real-time driver location update broadcast."""

    ride_id: str
    driver_id: str
    latitude: float
    longitude: float
    heading: Optional[float] = None
    speed: Optional[float] = None
    accuracy: Optional[float] = None
    recorded_at: str


class ETAPayload(BaseModel):
    """Computed ETA response including remaining distance and phase."""

    ride_id: str
    eta_minutes: Optional[float] = None
    current_eta_iso: Optional[str] = None
    remaining_distance_km: Optional[float] = None
    progress_percent: Optional[float] = None
    current_status: Optional[str] = None
    is_delayed: bool = False


class RouteProgressPayload(BaseModel):
    """Route progress snapshot for live map polyline updates."""

    ride_id: str
    progress_percent: float
    distance_traveled_km: float
    remaining_distance_km: float
    current_phase: str


class TrackingStartPayload(BaseModel):
    """Confirmation payload sent to room on tracking session start."""

    ride_id: str
    session_id: str
    driver_id: str
    status: str
    started_at: str
